[WIP] Unify case sensitive topic names

preetmishra opened the pull request:
This unifies case sensitive topic names using lowercase topic names as an invariant.

#### Crux
We need some kind of invariant which we can rely upon for lookups and comparisons given that topics can change their casing at any point. Consequently, I propose to use lowercase topics as keys wherever store/index data.

#### Commits
The current commit structure is temporary. I have fixed one thing per commit to represent how I went about the changes but we would definitely want to squash them (except the first) before merging.

I would greatly appreciate feedback about the proposal and what else should be fixed.

neiljp reviewed (commented):
@preetmishra This seems like a good first step (and the refactoring exposes a potential edge case bug?), but there various points that we need to address, as demonstrated by manual testing. This is not a complete list, but for example:
* If a user sends to `case` and `CASE` then two separate unread counts appear in the topic list, and only one appears in the message list if you're in that topic narrow
* I have an existing instance which triggered this where there are unreads in topics with two 'different' names (by case), both with unreads on czo, but your code is not combining them.
* Editing a message topic which matches by case causes it to disappear, but not appear in a new/different topic

Essentially, while fetching messages goes through `index_messages`, there are other situations we need to consider - anywhere where we compare topics, which is potentially a lot of different places.

Perhaps fundamentally, if we get an update to a topic (eg. edited latest message, or new messge), should we change the topic of every message we have stored? Or just update the rendering?

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -403,12 +403,14 @@ def index_messages(messages: List[Message],
-                topics_in_stream[msg['subject']] = set()
-            topics_in_stream[msg['subject']].add(msg['id'])
+        if msg['type'] == 'stream' and len(narrow) == 2:
+            narrow_topic = narrow[1][1]
```
The narrow can only have one topic, right?

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -403,12 +403,14 @@ def index_messages(messages: List[Message],
+            narrow_topic = narrow[1][1]
+            msg_topic = msg['subject']
+            if narrow_topic == msg_topic:
+                topics_in_stream = index['topic_msg_ids'][msg['stream_id']]
```
Discussed in [#zulip-terminal > Topic case sensitivity #T733](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Topic.20case.20sensitivity.20.23T733).

preetmishra commented:
@neiljp Thanks for the review and the pointers! :+1: 

I have reworked the fundamental approach that I had to now store lowercase topic names for lookups and comparisons (see https://github.com/zulip/zulip-terminal/pull/733#issue-455669765). I have also addressed the three issues that you reported.

sumanthvrao commented on zulipterminal/model.py:
```diff
@@ -775,7 +785,7 @@ def _handle_message_event(self, event: Event) -> None:
                     and (len(self.narrow) == 1
                          or (len(self.narrow) == 2
-                             and self.narrow[1][1] == message['subject']))):
+                             and self.narrow[1][1].lower() == message['subject'].lower()))):
```
It might be cleaner if we have a general helper of the kind `compare_lowercase` which can be used throughout our codebase to do lowercase comparisons.

preetmishra replied:
Thanks!

preetmishra committed:
refactor: helper: Simplify the topic index block in index_messages().

This extracts msg_topic and narrow_topics as variables and amends the
conditional accordingly.

preetmishra committed:
model: Use compare_lowercase() for topics while updating messages.

preetmishra committed:
conftest/helper/views: Use canonical_topic() for unread_counts.

Tests amended.

preetmishra committed:
helper/model: Update repr lookup for _have_last_message and pointer.

Also added narrow_with_canonical_topic().

preetmishra committed:
boxes: Unify case sensitive topic names for stream narrows.

preetmishra commented:
Updated with improved commits, more comments and test amendments (except which are related to muted topics).

preetmishra commented:
Updated to resolve conflicts.

preetmishra committed:
conftest/helper/model: Index topic_msg_ids using lowercase topic names.

The intent is to use lowercase topic names, as an invariant, in the data
structures that we locally use to keep track of topics and its metadata
(e.g. unread count).

canonicalize_topic() and compare_lowercase() are added as helpers.

Tests amended.

preetmishra committed:
model/views: Acknowledge case sensitive topics in topic list update.

Also, amended related _update_topic_index().

preetmishra committed:
model: Use canonical_topic() for muted topics.

neiljp commented:
@preetmishra This looks to cover a good number of comparison cases; this was blocked on another PR?

To clarify, this uses:
* lower case in internal structures (that seems clearest)
* "latest" topic names in display? (last-but-one commit references that?)
* the 'real' case in what remaining places? (for display ^)

I think this will be clearer with the now-merged #675 and when the topic list updates with something like #785, as we should be able to test internally more easily.

This looks good, though I've not dug into all the cases so far. Pending further review, this seems reasonable - my concern is whether we might consider locally handling topics with ids, which may simplify this issue - ie. each topic_id in a stream and (stream_id, topic_id) would be unique, and so we can have a 'latest name' (for display) for each id, and the comparisons would all occur at the point where topic names are converted to ids.

The first commit seems like a separate cleanup, is that correct?
